Add Soviet13a

Smittytron opened the pull request.

Smittytron committed:
Add Soviet13a

pchote commented:
Here's the output from my script decoder (still missing TeamType actions, I will try to come back to finishing this soon):

<details>
<summary>Click to expand</summary>

```miniyaml
Trigger@ENT1:
	On:
		 Attached cell entered or attached building captured by house "USSR"
	Actions:
		 Enable production for house "Good"
		 Run trigger "auto"
	Disable: After running once
	AttachedTo:
		Cells: 46,70, 47,70, 48,70, 49,70, 50,70, 46,71, 47,71, 48,71, 49,71, 50,71, 57,73, 58,73, 57,74, 58,74, 58,75, 59,75, 59,76, 60,76, 60,77, 61,77, 61,78, 62,78, 62,79, 63,79, 63,80, 64,80, 64,81, 65,81, 66,81, 67,81, 68,81, 69,81, 70,81, 71,81, 65,82, 66,82, 67,82, 68,82, 69,82, 70,82, 71,82
Trigger@LOS1:
	On:
		 House "Greece" owns building of type "DOME"
		 AND
		 Attached cell entered or attached building captured by house "USSR"
	Actions:
		 Set global scenario variable 1
		 Run trigger "los5"
	Disable: After running once
	AttachedTo:
		Cells: 23,99, 24,99, 25,99, 26,99, 27,99, 28,99, 29,99, 30,99, 31,99, 32,99, 23,100, 24,100, 25,100, 26,100, 27,100, 28,100, 29,100, 30,100, 31,100, 32,100, 23,101, 24,101, 25,101, 26,101, 29,101, 30,101, 31,101, 32,101, 23,102, 24,102, 25,102, 26,102, 29,102, 30,102, 31,102, 32,102, 23,103, 24,103, 25,103, 26,103, 27,103, 28,103, 29,103, 30,103, 31,103, 32,103, 23,104, 24,104, 25,104, 26,104, 27,104, 28,104, 29,104, 30,104, 31,104, 32,104, 23,105, 24,105, 25,105, 26,105, 27,105, 28,105, 29,105, 30,105, 31,105, 32,105, 23,106, 24,106, 25,106, 26,106, 27,106, 28,106, 29,106, 30,106, 31,106, 32,106
Trigger@hnt1:
	On:
		 0.0 minutes have elapsed
	Actions:
		 Create team "hunt1"
		 Reinforce with team "mcv1"
	Disable: After running once
Trigger@sea1:
	On:
		 35.0 minutes have elapsed
	Actions:
		 Create team "sea1"
	Disable: After running once
Trigger@prod:
	On:
		 0.0 minutes have elapsed
	Actions:
		 Enable production for house "Greece"
		 Enable production for house "Good"
	Disable: After running once
Trigger@ent2:
	On:
		 Attached cell entered or attached building captured by house "USSR"
	Actions:
		 Create team "hunt2"
		 Create team "hunt1"
	Disable: After running once
	AttachedTo:
		Cells: 101,57, 102,57, 103,57, 104,57, 105,57, 106,57, 107,57, 108,57, 109,57, 110,57, 111,57, 112,57, 113,57, 101,58, 102,58, 103,58, 104,58, 105,58, 106,58, 107,58, 108,58, 109,58, 110,58, 111,58, 112,58, 113,58, 101,59, 102,59, 103,59, 104,59, 105,59, 106,59, 107,59, 108,59, 109,59, 110,59, 111,59, 112,59, 113,59
Trigger@los2:
	On:
		 All buildings (excluding civilian) and units owned by house "USSR" are destroyed
	Actions:
		 House "USSR" is defeated
	Disable: After running once
Trigger@win1:
	On:
		 All buildings (excluding civilian) and units owned by house "Greece" are destroyed
		 AND
		 All buildings (excluding civilian) and units owned by house "Good" are destroyed
	Actions:
		 House "USSR" is victorious
	Disable: After running once
Trigger@chro:
	On:
		 Attached unit or building is destroyed
		 LINKED
		 Attached cell entered or attached building captured by house "USSR"
	Actions:
		 House "USSR" is defeated
		 House "USSR" is victorious
	Disable: After running once
Trigger@los4:
	On:
		 Global scenario variable 1 is set
		 LINKED
		 Attached cell entered or attached building captured by house "USSR"
	Actions:
		 Destroy attached buildings
		 Run trigger "win2"
	Disable: After running once
	AttachedTo:
		 Actor1 (pdox owned by house "Germany" at cell 27,101)
Trigger@win2:
	On:
		 Manual Trigger
	Actions:
		 House "USSR" is victorious
		 Play speech message "OBJRCH1"
	Disable: After running once
Trigger@win3:
	On:
		 Attached unit or building is destroyed
	Actions:
		 Disable trigger "ENT1"
	Disable: After running on all attached objects
	AttachedTo:
		 Actor10 (dome owned by house "Greece" at cell 44,104)
		 Actor37 (dome owned by house "Greece" at cell 29,90)
		 Actor38 (dome owned by house "Greece" at cell 48,88)
		 Actor39 (dome owned by house "Greece" at cell 27,75)
Trigger@reac:
	On:
		 Attached unit or building is destroyed
	Actions:
		 Destroy attached buildings
	Disable: After running once
	AttachedTo:
		 Actor47 (brl3 owned by house "Greece" at cell 70,90)
		 Actor51 (v19 owned by house "Greece" at cell 67,89)
		 Actor52 (v19 owned by house "Greece" at cell 69,91)
		 Actor68 (v19 owned by house "Greece" at cell 26,76)
		 Actor69 (brl3 owned by house "Greece" at cell 26,77)
		 Actor72 (v19 owned by house "Greece" at cell 52,101)
		 Actor73 (brl3 owned by house "Greece" at cell 52,102)
Trigger@air1:
	On:
		 21.0 minutes have elapsed
		 AND
		 Global scenario variable 2 is set
	Actions:
		 Reinforce with team "air1"
	Disable: After running once
Trigger@air2:
	On:
		 32.0 minutes have elapsed
	Actions:
		 Set global scenario variable 2
	Disable: After running once
Trigger@air3:
	On:
		 Attached unit or building is destroyed
	Actions:
		 Clear global scenario variable 2
	Disable: After running on all attached objects
	AttachedTo:
		 Actor26 (hpad owned by house "Greece" at cell 32,73)
		 Actor27 (hpad owned by house "Greece" at cell 49,100)
		 Actor40 (hpad owned by house "Greece" at cell 56,76)
		 Actor54 (hpad owned by house "Greece" at cell 61,80)
Trigger@mly1:
	On:
		 0.1 minutes have elapsed
	Actions:
		 Create team "mlay1"
	Disable: After running once
Trigger@mly2:
	On:
		 3.0 minutes have elapsed
	Actions:
		 Create team "mlay1"
	Disable: After running once
Trigger@para:
	On:
		 House "USSR" has built building of type "AFLD"
	Actions:
		 Enable repeating superweapon "PARA_BOMB" for house "USSR"
		 Enable repeating superweapon "PARA_INFANTRY" for house "USSR"
	Disable: After running once
Trigger@auto:
	On:
		 15.0 minutes have elapsed
	Actions:
		 Enable AI autocreate for house "Greece"
		 Enable AI autocreate for house "Good"
	Disable: After running once
Trigger@glbl:
	On:
		 0.0 minutes have elapsed
	Actions:
		 Set global scenario variable 5
		 Set global scenario variable 6
	Disable: After running once
Trigger@dst1:
	On:
		 Attached unit or building is destroyed
	Actions:
		 Clear global scenario variable 5
	Disable: After running once
	AttachedTo:
		 Actor41 (syrd owned by house "Greece" at cell 73,84)
Trigger@dst2:
	On:
		 Attached unit or building is destroyed
	Actions:
		 Clear global scenario variable 6
	Disable: After running once
Trigger@lst1:
	On:
		 Global scenario variable 3 is set
	Actions:
		 Create team "lst9"
	Disable: After running once
Trigger@lst2:
	On:
		 6.0 minutes have elapsed
	Actions:
		 Reinforce with team "lst5"
		 Reinforce with team "lst3"
	Disable: After running once
Trigger@lst3:
	On:
		 13.0 minutes have elapsed
		 AND
		 Global scenario variable 5 is set
	Actions:
		 Reinforce with team "lst2"
	Disable: After running once
Trigger@lst4:
	On:
		 23.0 minutes have elapsed
		 AND
		 Global scenario variable 6 is set
	Actions:
		 Reinforce with team "lst6"
	Disable: After running once
Trigger@lst5:
	On:
		 34.0 minutes have elapsed
		 AND
		 Global scenario variable 7 is set
	Actions:
		 Reinforce with team "lst1"
		 Reinforce with team "lst7"
	Disable: After running once
Trigger@set7:
	On:
		 0.0 minutes have elapsed
	Actions:
		 Set global scenario variable 7
	Disable: After running once
Trigger@clr7:
	On:
		 Global scenario variable 5 is not set
		 AND
		 Global scenario variable 6 is not set
	Actions:
		 Clear global scenario variable 7
	Disable: After running once
Trigger@chn1:
	On:
		 31.0 minutes have elapsed
	Actions:
		 Reinforce with team "chin1"
	Disable: After running once
Trigger@lst6:
	On:
		 63.5 minutes have elapsed
		 AND
		 Global scenario variable 7 is set
	Actions:
		 Reinforce with team "lst2"
		 Reinforce with team "lst7"
	Disable: After running once
Trigger@lst7:
	On:
		 50.0 minutes have elapsed
		 AND
		 Global scenario variable 5 is set
	Actions:
		 Reinforce with team "lst1"
		 Reinforce with team "lst3"
	Disable: After running once
Trigger@lst8:
	On:
		 72.0 minutes have elapsed
		 AND
		 Global scenario variable 6 is set
	Actions:
		 Reinforce with team "lst6"
		 Reinforce with team "chin1"
	Disable: After running once
Trigger@lst9:
	On:
		 84.0 minutes have elapsed
		 AND
		 Global scenario variable 5 is set
	Actions:
		 Reinforce with team "lst1"
		 Reinforce with team "lst7"
	Disable: After running once
Trigger@chn2:
	On:
		 63.2 minutes have elapsed
	Actions:
		 Reinforce with team "chin1"
	Disable: After running once
Trigger@air4:
	On:
		 36.2 minutes have elapsed
		 AND
		 Global scenario variable 2 is set
	Actions:
		 Reinforce with team "air2"
	Disable: After running once
Trigger@air5:
	On:
		 63.0 minutes have elapsed
		 AND
		 Global scenario variable 2 is set
	Actions:
		 Reinforce with team "air3"
		 Reinforce with team "hunt1"
	Disable: Never
Trigger@air6:
	On:
		 Attached unit or building is destroyed
	Actions:
		 Reinforce with team "air4"
	Disable: After running on all attached objects
	AttachedTo:
		 TeamType "air3"
Trigger@los3:
	On:
		 All buildings (excluding civilian) owned by house "Germany" are destroyed
	Actions:
		 House "USSR" is defeated
	Disable: After running once
Trigger@los5:
	On:
		 Manual Trigger
	Actions:
		 Reveal area around "waypoint75" (27,102)
		 Display text message 55 from tutorial.ini
	Disable: After running once
Trigger@sea2:
	On:
		 95.0 minutes have elapsed
		 AND
		 Global scenario variable 5 is set
	Actions:
		 Create team "sea2"
	Disable: After running once
TeamType@hunt1:
	Units: 2TNK, 2TNK, 2TNK, ARTY, ARTY, MGG
TeamType@sea1:
	Units: CA
TeamType@mcv1:
	Units: MCV
TeamType@sea2:
	Units: DD, DD, CA
TeamType@hunt2:
	Units: 2TNK, 2TNK, 2TNK, 2TNK, ARTY, ARTY, ARTY
TeamType@air1:
	Units: HELI, HELI, HELI, HELI
TeamType@auto1:
	Units: 2TNK, 2TNK, 1TNK
TeamType@auto2:
	Units: E3, E3, E3, ARTY, ARTY, ARTY
TeamType@auto3:
	Units: 1TNK, 1TNK, 1TNK, JEEP, JEEP
TeamType@auto4:
	Units: JEEP, JEEP, JEEP
TeamType@mlay1:
	Units: MNLY
TeamType@auto5:
	Units: 2TNK, ARTY, ARTY
TeamType@auto6:
	Units: ARTY, ARTY, ARTY, MGG
TeamType@lst1:
	Units: 1TNK, ARTY, ARTY, LST
TeamType@lst2:
	Units: 1TNK, 1TNK, 1TNK, LST
TeamType@lst3:
	Units: E3, E3, E3, E3, E3, LST
TeamType@lst4:
	Units: ARTY, ARTY, ARTY, ARTY, MGG, LST
TeamType@lst5:
	Units: E3, E3, E3, E3, MEDI, LST
TeamType@lst6:
	Units: 2TNK, 2TNK, ARTY, MGG
TeamType@lst7:
	Units: 2TNK, 2TNK, 2TNK, 2TNK, ARTY, LST
TeamType@chin1:
	Units: E3, E3, E3, E3, MEDI, TRAN
TeamType@auto7:
	Units: 2TNK, 2TNK, ARTY, ARTY, MGG
TeamType@auto8:
	Units: 1TNK, ARTY, ARTY, ARTY, MGG
TeamType@auto9:
	Units: 2TNK, 2TNK, 2TNK, 1TNK, 1TNK
TeamType@lst8:
	Units: MNLY, LST
TeamType@lst9:
	Units: MNLY, LST
TeamType@air2:
	Units: HELI, HELI, HELI, HELI
TeamType@air3:
	Units: HELI, HELI, HELI, HELI
	Trigger: air6
TeamType@air4:
	Units: HELI, HELI, HELI, HELI
ScriptedBuildingProduction: house "GoodGuy"
	GUN at cell 63,78
	GUN at cell 59,74
	GUN at cell 46,69
	GUN at cell 50,69
```
</summary>

tttppp commented:
Trying this on easy I failed twice. First attempt I was outflanked by the tanks that seem to come from behind and lost my MCV before I'd even got a barracks.

Second time I managed to grind long enough to stabilise (at around 18 mins in) and cleaned up most of the first base by 22 mins.  However I then pushed too far towards the south and immediately lost due to getting too near to the chronosphere.  I didn't actually know where the chronosphere was, so it was quite hard to stay away from it - particularly as there was a radar dome fairly near by.

tttppp commented:
I failed this twice more on easy. First time my army and MCV was destroyed by the artis and medium tanks at the start.  Second time I lost my MCV to artis trying to set up base next to the ore patch across the river to the south.

Smittytron commented:
> Second time I managed to grind long enough to stabilise (at around 18 mins in) and cleaned up most of the first base by 22 mins. However I then pushed too far towards the south and immediately lost due to getting too near to the chronosphere. I didn't actually know where the chronosphere was, so it was quite hard to stay away from it - particularly as there was a radar dome fairly near by.

I added a camera to the Chronosphere location and a message explaining the trap wasn't disarmed when the player triggers it.

abcdefg30 commented on mods/ra/maps/soviet-13a/soviet13a-AI.lua:
```diff
@@ -0,0 +1,175 @@
+		if not unit.IsDead then
+			Trigger.OnIdle(unit, unit.Hunt)
+		end
+	end)
```
```suggestion
	Utils.Do(AttackGroup, IdleHunt)
```
?

abcdefg30 commented on mods/ra/maps/soviet-13a/soviet13a-AI.lua:
```diff
@@ -0,0 +1,175 @@
+	Greece.Build({ Utils.Random(AlliedInfantry) }, function(units)
+		table.insert(AttackGroup, units[1])
+		SendAttackGroup()
+		Trigger.AfterDelay(ProductionInterval[Map.LobbyOption("difficulty")], ProduceInfantry)
```
Can we pull `Map.LobbyOption("difficulty")` out into a global variable so we can move that call more easily to the global script as soon as all missions have a difficulty? (You already call this in `ActivateAI`, just make the variable there a global `Difficulty` variable.)

Smittytron commented:
Updated this and Soviet13b.

abcdefg30 reviewed (dismissed):
There is a tile error in the lower left corner (not entirely the bottom, it is between the left two radars):
![grafik](https://user-images.githubusercontent.com/7704140/138953590-d33b6df6-a18d-4004-b27e-9cecdb8c59ee.png)

Can you also add a slight delay to the "Mission Accomplished" speech, so that it doesn't overlap with "Structure Captured"?

Smittytron commented:
Fixed tile error and extended announcement delay.

Orb370 commented:
I played through it on hard. I found it a little easy, considering I beat it the first time no sweat. I'd recommend starting it with less V2's so you have to rely more on your mammoth regen.

I assumed I had to take out two radar domes based on the wording (the link between them?), but I think that's not the case given how many radar domes were on the map? I would change the wording to "Destroy all Allied Radar Sites before..."

There's also a typo for the objectives: "Capture the Chronoshpere". Chronosphere is misspelled in both.

Replay: [Orb-Hard-13a.zip](https://github.com/OpenRA/OpenRA/files/7455560/Orb-Hard-13a.zip)

Smittytron commented:
Updated for #19738

PunkPun commented:
Needs to be rebased and translations added #20387

Smittytron commented:
Updated for translations

PunkPun commented:
Needs a rebase and difficulty translations added. See changelog #20585 for how.

Smittytron commented:
Updated for #20585

abcdefg30 commented on mods/ra/maps/soviet-13a/soviet13a.lua:
```diff
@@ -0,0 +1,111 @@
+	end)
+
+	Trigger.OnCapture(Chronosphere, function()
+		USSR.MarkCompletedObjective(CaptureChronosphere)
```
This should fail when the radar domes are not yet captured.

JovialFeline replied:
[There are some trigger cells in the road](https://i.imgur.com/KUr5zLs.png) intended to self-destruct the sphere before that could be an issue. However, it's possible to capture a helipad, pilot a transport around that, and drop off an engineer. Loading grenadiers for the wall is optional since the paratroopers include a pair.

It's not quite a skip since the domes are still needed and the next attack wave will destroy the sphere if a backdoor capture succeeds, but I doubt that order was intended.

So yeah, probably.

Smittytron replied:
Chronosphere will now detonate on capture if radars aren't destroyed.

abcdefg30 reviewed (commented):
Lgtm otherwise.

JovialFeline commented:
I ran through a few times and noted some things:

- GoodGuy is not defined. When SendCruiser checks GoodguyShipyard's owner, it compares to nil and cruisers are never sent.
- Rather than Greece's team near the bridge, the opening attack is done by the southeastern team. The +2 medium tanks and +1 artillery may be why the other commenter had some early difficulty. I think I still prefer this to the SE team feeling like punishment for early scouting/mining where a trigger happens to be.
- The airlifted soldiers can sometimes get stuck on trees. This is less to do with the script and more of a pathfinding quirk with Hunt. This behavior is not the most consistent, and they can get unstuck if they're disturbed in some way. [Here, the rocket soldiers wiggle loose to target yaks, while the rifles stay put.](https://i.imgur.com/mJFHtye.mp4) If that's a bother, attack-moving them south for two cells before Hunt may work.
- The Chronosphere team was a pleasant surprise!

Smittytron commented:
Addressed @JovialFeline's points: 
- Added GoodGuy as a player
- Swapped bridge group to start attack and the south group to attack on footprint trigger.
- Moved the Chinook LZ a few spaces.

JovialFeline commented:
I did notice some different [stalled hunters](https://imgur.com/a/vNZ1vVy), but I'm more confused by Hunt's behavior this time. I'm aware the original likewise dumps them off with a simple Hunt order.

Beyond that, everything seemed smooth. Neat touch on the ore footprint.

PunkPun commented on mods/ra/maps/soviet-13a/soviet13a.lua:
```diff
@@ -0,0 +1,119 @@
+
+	Trigger.OnCapture(Chronosphere, function()
+		if not USSR.IsObjectiveCompleted(TakeDownRadar) then
+			Chronosphere.Kill()
```
```suggestion
			Media.DisplayMessage(UserInterface.Translate("chrono-trap-triggered"), UserInterface.Translate("headquarters"))
			Chronosphere.Kill()
```

this should also trigger the message

PunkPun reviewed (commented):
Otherwise LGTM

I didn't encounter the hunt glitch, the only issue I had is idling longbows. Often they weren't attacking nor rearming.

Overall this feels like a fairly easy and short mission. You can just use V2's to take out the behind the cliff radars, perhaps there needs to be something extra on hard?

Also should we consider removing GPS from campaign?

PunkPun commented:
IMO GPS should be disabled for this mission. The AI doesn't really care about it and its existence might only confuse players. At least I thought it was an important timer when playing the mission

Smittytron commented:
Changes made from feedback here and on Discord:

- Added "chrono-trap-triggered" message to new kill trigger.
- GPS timer relationship changed to Ally so it won't appear to the player unless captured.
- AlliedVehiclesUpgradeDelay changed from 18 to 8 minutes.
-  Added artillery by south and southwestern radar domes.
- Added a rally point for the enemy WarFactory because tanks were getting stuck.
- Player won't start with V2's on Hard difficulty.

PunkPun reviewed (dismissed):
LGTM

Mailaender commented on mods/ra/maps/soviet-13a/soviet13a-AI.lua:
```diff
@@ -0,0 +1,170 @@
+
+ActivateAI = function()
+	WTransUnits = WTransUnits[Difficulty]
+	WTransDelays = WTransDelays[Difficulty]
```
```suggestion
	WTransDelay = WTransDelays[Difficulty]
```
would make more sense and silence the linter.

JovialFeline commented on mods/ra/maps/soviet-13a/map.yaml:
```diff
@@ -0,0 +1,939 @@
+		NonCombatant: True
+		Faction: england
+	PlayerReference@Greece:
+		Name: Greece
```
I won't pretend to fully understand why, but giving Greece a bot seems to fix the idle helicopters that @PunkPun  mentioned. Without it, Longbows can skip the usual ReturnToBase after using up their ammo pool, and they'll just hover.

Other than that, I ran through it twice today without any issues coming from the script.
```suggestion
		Name: Greece
		Bot: campaign
```

PunkPun replied:
Do we do this in any of the other missions though?

JovialFeline replied:
Soviet-01, Soviet-02a, Soviet-02b, Soviet-03, Soviet-04a, Soviet-04b, Soviet-05 (with a tweaked skirmish bot for the island), Soviet-06a, Soviet-06b, Soviet-07, Soviet-08a, Soviet-08b, Soviet-09, and Soviet-10 all seem to do this. It seems to be the norm in the Allies' missions as well.

PunkPun replied:
Interesting, did you test for any potential unwanted side effects?

JovialFeline replied:
I did run through the map a third time without any apparent issues from the campaign bot, after I had put together a small test map to confirm the different behavior. While I was at it, I also used the test map to see if the clogged factory exit was related, and it didn't seem to be.

Smittytron commented:
Updated

Mailaender commented on mods/ra/maps/soviet-13a/soviet13a-AI.lua:
```diff
@@ -0,0 +1,170 @@
+	easy = { { "2tnk", "e1", "e1", "e3", "e3" }, { "1tnk", "1tnk", "jeep", "jeep" } }
+}
+
+WTransDelay =
```
```suggestion
WTransDelays =
```

You misunderstood. This should stay plural. Don't overwrite the table with a variable.

Smittytron commented:
Updated

PunkPun commented on mods/ra/maps/soviet-13a/soviet13a-AI.lua:
```diff
@@ -0,0 +1,170 @@
+{
+	easy = DateTime.Minutes(5),
+	normal = DateTime.Minutes(4),
+	hard = DateTime.Seconds(3)
```
```suggestion
	hard = DateTime.Minutes(3)
```

This change definitely needs to be reverted 😅

PunkPun reviewed (commented):
would it be worth to kill the artilleries on easier difficulties?

Smittytron commented:
Updated

> would it be worth to kill the artilleries on easier difficulties?

I think they'll be fine. It will be at the end of the mission where the player is already built up no matter what their skill level is.

abcdefg30 commented on mods/ra/maps/soviet-13a/soviet13a.lua:
```diff
@@ -0,0 +1,125 @@
+	MissionTriggers()
+	ActivateAI()
+	GreeceWarFactory.RallyPoint = WarFactoryRally.Location
+	if Map.LobbyOption("difficulty") == "hard" then
```
Can we use the global variable here?
```suggestion
	if Difficulty == "hard" then
```

PunkPun replied:
force pushed

abcdefg30 reviewed (commented):
Code changes lgtm otherwise.

PunkPun reviewed (approved):
LGTM

PunkPun commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a0cc1c4bd0dc4c708d80e6416b2f16e24864c534...7b67f43b248195e5f912fc25981f3b08d841df33)
